src/log.h

```diff
@@ -185,35 +188,155 @@ class LogOutputBuffer : public ICombinedLogOutput {
-extern std::ostream null_stream;
 
-extern std::ostream *dout_con_ptr;
-extern std::ostream *derr_con_ptr;
-extern std::ostream *derr_server_ptr;
+class StreamProxy {
+public:
+	StreamProxy(std::ostream *os) : m_os(os) { }
+
+	template<typename T>
+	StreamProxy& operator<<(T&& arg) {
+		if (m_os) {
+			*m_os << std::forward<T>(arg);
+		}
+		return *this;
+	}
 
-extern Logger g_logger;
+	StreamProxy& operator<<(std::ostream& (*manip)(std::ostream&)) {
+		if (m_os) {
+			*m_os << manip;
+		}
+		return *this;
+	}
 
-// Writes directly to all LL_NONE log outputs for g_logger with no prefix.
-extern std::ostream rawstream;
+private:
+	std::ostream *m_os;
+};
 
-extern std::ostream errorstream;
-extern std::ostream warningstream;
-extern std::ostream actionstream;
-extern std::ostream infostream;
-extern std::ostream verbosestream;
-extern std::ostream dstream;
 
-#define TRACEDO(x) do {               \
-	if (g_logger.getTraceEnabled()) { \
-		x;                            \
-	}                                 \
-} while (0)
+class LogStream {
+public:
+	LogStream() = delete;
+	LogStream(const LogStream &) = delete;
+
+	LogStream(LogTarget &target) :
+		m_target(target),
+		m_buffer(*this),
+		m_dummy_buffer(),
+		m_stream(&m_buffer),
+		m_dummy_stream(&m_dummy_buffer),
+		m_proxy(&m_stream),
+		m_dummy_proxy(nullptr) { }
+
+	template<typename T>
+	StreamProxy& operator<<(T&& arg) {
+		StreamProxy& sp = m_target.hasOutput() ? m_proxy : m_dummy_proxy;
+		sp << std::forward<T>(arg);
+		return sp;
+	}
 
-#define TRACESTREAM(x) TRACEDO(verbosestream x)
+	StreamProxy& operator<<(std::ostream& (*manip)(std::ostream&)) {
+		StreamProxy& sp = m_target.hasOutput() ? m_proxy : m_dummy_proxy;
+		sp << manip;
+		return sp;
+	}
 
-#define dout_con (*dout_con_ptr)
-#define derr_con (*derr_con_ptr)
+	operator bool() {
+		return m_target.hasOutput();
+	}
 
+	void internalFlush(const std::string &buf) {
+		m_target.log(buf);
+	}
+
+	operator std::ostream&() {
+		return m_target.hasOutput() ? m_stream : m_dummy_stream;
+	}
+
+private:
+	class StringBuffer : public std::streambuf {
+	public:
+		StringBuffer(LogStream &p) : parent(p) {
+			buffer_index = 0;
+		}
+
+		int overflow(int c) {
+			push_back(c);
+			return c;
+		}
+
+		void push_back(char c) {
+			if (c == '\n' || c == '\r') {
+				if (buffer_index)
+					parent.internalFlush(std::string(buffer, buffer_index));
+				buffer_index = 0;
+			} else {
+				buffer[buffer_index++] = c;
+				if (buffer_index >= BUFFER_LENGTH) {
+					parent.internalFlush(std::string(buffer, buffer_index));
+					buffer_index = 0;
+				}
+			}
+                }
+
+	        std::streamsize xsputn(const char *s, std::streamsize n) {
+			for (int i = 0; i < n; ++i)
+				push_back(s[i]);
+			return n;
+		}
+
+	private:
+		// 10 streams per thread x (256 + overhead) ~ 3K per thread
+		static const int BUFFER_LENGTH = 256;
+		LogStream &parent;
+		char buffer[BUFFER_LENGTH];
+	        int buffer_index;
+	};
```

Comment:
Why inline the implementations of this class?

Reply:
I put it there to emphasize that it is used internally to LogStream and not externally visible.

I could move it out with a bit of work. It holds a reference to LogStream, so if I pulled it out I'd need to make LogStream virtual or make it templated.